### PR TITLE
Show expected elimination forms on optic kind mismatch

### PR DESCRIPTION
**Pull request by arybczak**
```haskell
λ> view (_1 % _Left)

<interactive>:163:1: error:
    • An_AffineTraversal cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘(^?)’ (from Optics.Operators)
        ‘over’ (from Optics.Setter)
        ‘(%~)’ (from Optics.Operators)
        ‘set’ (from Optics.Setter)
        ‘(.~)’ (from Optics.Operators)
    • In the expression: view (_1 % _Left)
      In an equation for ‘it’: it = view (_1 % _Left)
λ> view traversed

<interactive>:164:1: error:
    • A_Traversal cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘traverseOf’ (from Optics.Traversal)
        ‘traverseOf_’ (from Optics.Fold)
        ‘foldMapOf’ (from Optics.Fold)
        ‘toListOf’ (from Optics.Fold)
        ‘(^..)’ (from Optics.Operators)
    • In the expression: view traversed
      In an equation for ‘it’: it = view traversed
λ> view _Left

<interactive>:165:1: error:
    • A_Prism cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘(^?)’ (from Optics.Operators)
        ‘review’ (from Optics.Review)
        ‘(#)’ (from Optics.Operators)
    • In the expression: view _Left
      In an equation for ‘it’: it = view _Left
λ> review _1

<interactive>:166:1: error:
    • A_Lens cannot be used as A_Review
      Perhaps you meant to use one of these:
        ‘view’ (from Optics.Getter)
        ‘(^.)’ (from Optics.Operators)
        ‘over’ (from Optics.Setter)
        ‘(%~)’ (from Optics.Operators)
        ‘set’ (from Optics.Setter)
        ‘(.~)’ (from Optics.Operators)
    • In the expression: review _1
      In an equation for ‘it’: it = review _1
λ> traverseOf_ mapped

<interactive>:167:1: error:
    • A_Setter cannot be used as A_Fold
      Perhaps you meant to use one of these:
        ‘over’ (from Optics.Setter)
        ‘(%~)’ (from Optics.Operators)
        ‘set’ (from Optics.Setter)
        ‘(.~)’ (from Optics.Operators)
    • In the expression: traverseOf_ mapped
      In an equation for ‘it’: it = traverseOf_ mapped
```

🙂

### CODE REVIEW

File: optics-core/src/Optics/Internal/Optic/Subtyping.hs [hunk 1]
```diff
@@ -26,17 +27,71 @@ class Is k l where
...
+  EliminationForms An_AffineTraversal = AffineTraversalEliminations
+  EliminationForms A_Fold             = FoldEliminations
+  EliminationForms A_Getter           = GetterEliminations
+  EliminationForms A_Lens             = LensEliminations
+  EliminationForms A_Prism            = PrismEliminations
+  EliminationForms A_ReversedLens     = PrismEliminations
+  EliminationForms A_ReversedPrism    = LensEliminations
+  EliminationForms A_Review           = ReviewEliminations
+  EliminationForms A_Setter           = SetterEliminations
+  EliminationForms A_Traversal        = TraversalEliminations
```
Comment (phadej) [inline optics-core/src/Optics/Internal/Optic/Subtyping.hs:53]: This isn't total, it is missing `An_Iso`. I guess one doesn't need it as `Iso` can be used as anything, but I'd like `type family` to be total anyway.

File: optics-core/src/Optics/Internal/Optic/Subtyping.hs [hunk 2]
```diff
@@ -26,17 +27,71 @@ class Is k l where
...
+  =     AffineFoldEliminations
+  ':$$: ReviewEliminations
+
+type ReviewEliminations
+  =     ShowSymbolWithOrigin "review" "Optics.Review"
+  ':$$: ShowSymbolWithOrigin "(#)"    "Optics.Operators"
+
+type SetterEliminations
+  =     ShowSymbolWithOrigin "over" "Optics.Setter"
+  ':$$: ShowSymbolWithOrigin "(%~)" "Optics.Operators"
```
Comment (phadej) [inline optics-core/src/Optics/Internal/Optic/Subtyping.hs:87]: I find it's slightly messy, would be nicer to have named combinators as they are now, but all operators as a single list.

I.e not

```
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘(^?)’ (from Optics.Operators)
        ‘over’ (from Optics.Setter)
        ‘(%~)’ (from Optics.Operators)
        ‘set’ (from Optics.Setter)
        ‘(.~)’ (from Optics.Operators)
```

but

```
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘over’ (from Optics.Setter)
        ‘set’ (from Optics.Setter)
        ‘(^?)’ ‘(%~)’ ‘(.~)’ (from Optics.Operators)
```

How hard would to do that change?

(I wouldn't group named combinators by module, but that's an option too if it makes stuff easier).

**Comment by arybczak** [comment]
Done, now it looks like this:

```haskell
λ> view (_1 % _Left)

<interactive>:20:1: error:
    • An_AffineTraversal cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘over’ (from Optics.Setter)
        ‘set’ (from Optics.Setter)
        ‘(^?)’ ‘(%~)’ ‘(.~)’ (from Optics.Operators)
    • In the expression: view (_1 % _Left)
      In an equation for ‘it’: it = view (_1 % _Left)
λ> view traversed

<interactive>:21:1: error:
    • A_Traversal cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘traverseOf’ (from Optics.Traversal)
        ‘traverseOf_’ (from Optics.Fold)
        ‘foldMapOf’ (from Optics.Fold)
        ‘toListOf’ (from Optics.Fold)
        ‘(^..)’ (from Optics.Operators)
    • In the expression: view traversed
      In an equation for ‘it’: it = view traversed
λ> view _Left

<interactive>:22:1: error:
    • A_Prism cannot be used as A_Getter
      Perhaps you meant to use one of these:
        ‘preview’ (from Optics.AffineFold)
        ‘review’ (from Optics.Review)
        ‘(^?)’ ‘(#)’ (from Optics.Operators)
    • In the expression: view _Left
      In an equation for ‘it’: it = view _Left
λ> review _1

<interactive>:23:1: error:
    • A_Lens cannot be used as A_Review
      Perhaps you meant to use one of these:
        ‘view’ (from Optics.Getter)
        ‘over’ (from Optics.Setter)
        ‘set’ (from Optics.Setter)
        ‘(^.)’ ‘(%~)’ ‘(.~)’ (from Optics.Operators)
    • In the expression: review _1
      In an equation for ‘it’: it = review _1
λ> traverseOf_ mapped

<interactive>:24:1: error:
    • A_Setter cannot be used as A_Fold
      Perhaps you meant to use one of these:
        ‘over’ (from Optics.Setter)
        ‘set’ (from Optics.Setter)
        ‘(%~)’ ‘(.~)’ (from Optics.Operators)
    • In the expression: traverseOf_ mapped
      In an equation for ‘it’: it = traverseOf_ mapped
```

File: optics-core/src/Optics/Internal/Optic/Subtyping.hs [hunk 3]
```diff
@@ -26,17 +27,68 @@ class Is k l where
...
-  implies _ = error "unreachable"
-
 -- | Every kind of optic can be used as itself.
 instance Is k k where
   implies r = r
 
+-- | Overlappable instance for a custom type error.
+instance {-# OVERLAPPABLE #-} TypeError
+  ('ShowType k ':<>: 'Text " cannot be used as " ':<>: 'ShowType l
+   ':$$: 'Text "Perhaps you meant to use one of these:"
```
Comment (adamgundry) [inline optics-core/src/Optics/Internal/Optic/Subtyping.hs:37]: Very neat idea. In general I'm a bit cautious about making error messages too long, as GHC error messages are already rather verbose IMHO, but in this case I think it's worth it.

Nitpick: the verb "use" appears in two different senses, first as "using an optic kind at a superclass" then as "using a function to eliminate an optic". How about "apply" in the second case?
```suggestion
   ':$$: 'Text "Perhaps you meant to apply one of these:"
```

Comment (arybczak) [reply optics-core/src/Optics/Internal/Optic/Subtyping.hs:37]: You're right about double `use`, but apply feels weird to me...

What about simply `Perhaps you meant one of these:` instead? This is what GHC gives you when you misspell things.

```haskell
λ> view_ _Left

<interactive>:35:1: error:
    • Variable not in scope:
        view_ :: Prism (Either a0 b0) (Either c0 b0) a0 c0 -> t
    • Perhaps you meant one of these:
        ‘view’ (imported from Optics.Getter),
        ‘views’ (imported from Optics.Getter)
```

Comment (adamgundry) [reply optics-core/src/Optics/Internal/Optic/Subtyping.hs:37]: Yes, that seems okay to me. With "apply" I was trying to hint that the suggestions were for replacing the function being applied to the optic, but I agree it feels a bit odd.

**Comment by arybczak** [comment]
All suggestion applied.

**Comment by kosmikus** [comment]
This looks nice. Looking forward to it.

**Pull request merged**